docs/en/guides/90-community/02-rfcs/20240721-external-dictionary.md

```diff
@@ -0,0 +1,205 @@
+pub struct DictionaryAsyncFunction {
+    engine: DictionarySourceEngine,
+    // dictonary address, for examaple: mysql://root:123@0.0.0.0:3306/default
+    url: String,
+    // sql to get the value from source table.
+    // for example: select name from user_info where id=1;
+    query_sql: String,
+    return_type: DataType,
+    //Specify the maximum time to attempt a connection to the data source.
+    connection_timeout: std::time::Duration,
+    //Specify the maximum execution time for the query operation.
+    query_timeout: std::time::Duration,
+    //Used to store additional parameters that the query might require, such as the values for placeholders in the SQL query.
+    params: Vec<ParameterValue>,
+}
+```
+
+Rename `AsyncFunction` in the `async_function` module to `AsyncFunctionDesc` to avoid naming conflicts with the logical and physical plan of AsyncFunction. Additionally, include `DictionaryAsyncFunction`. The definition is as follows:
+
+```rust
+pub enum AsyncFunctionDesc {
+     SequenceAsyncFunction(SequenceAsyncFunction),
+     DictonaryAsyncFunction(DictionaryAsyncFunction),
+}
+```
+
+Update the `AsyncFunction` definition in both the logical and physical plans by adding the `AsyncFunctionDesc` field. This process reuses existing logic for generating dictionary AsyncFunction logical and physical plans.
+
+- The struct of logical plan is as follows:
+
+```rust
+pub struct AsyncFunction {
+   pub func_name: String,
+   pub display_name: String,
+   pub arguments: Vec<String>,
+   pub return_type: DataType,
+   pub index: IndexType,
+   pub desc: AsyncFunctionDesc,//Newly added property
+}
+```
+
+- The struct of physical plan is as follows:
+
+```rust
+pub struct AsyncFunction {
+   pub plan_id: u32,
+   pub func_name: String,
+   pub display_name: String,
+   pub arguments: Vec<String>,
+   pub return_type: DataType,
+   pub schema: DataSchemaRef,
+   pub input: Box<PhysicalPlan>,
+   pub stat_info: Option<PlanStatsInfo>,
+   pub desc: AsyncFunctionDesc,//Newly added property
+}
+```
+
+The `Transform` in the pipeline, where the actual reading of external data takes place, can be defined as follows:
+
+```rust
+pub struct TransformDictionary {
+     ctx: Arc<QueryContext>,
+     dict_func: DictionaryAsyncFunction,
+}
+```
+
+Implement the `transform` method of the `AsyncTransform` trait and call an external database to obtain dictionary data. The main process is illustrated in the following diagram:
+
+<img src="/docs/public/img/rfc/20240721-external-dictionary/external-dictionary-1.png" alt="Flowchart of getting external data" style={{zoom:"80%"}} />
+
+The execution process of the `dict_get` function is summarized in the following diagram:
+
+<img src="/docs/public/img/rfc/20240721-external-dictionary/external-dictionary-2.png" alt="Flowchart of the dict_get" style={{zoom:"80%"}} />
+
+## Unresolved questions
+
+- In addition to supporting Protobuf file definitions, should XML file configuration also be supported?
```

Comment:
We don't support XML config, maybe we can add builtin dictionary in config in future.